hrl/agent/dsc/classifier/sift_classifier.py

```diff
@@ -0,0 +1,322 @@
+
+    def fit(self, X=None, Y=None, svm_type='svc', gamma='scale', nu=0.1):
+        """
+        train the classifier in 3 steps
+        1. train the kmeans classifier based on the SIFT features of images
+        2. extract the histogram of the SIFT features
+        3. train an SVM classifier based on the histogram features
+        args:
+            x: a list of images
+            y: a list of labels (str)
+        """
+        assert svm_type in ['svc', 'one_class_svm']
+        X = list(X)  # in case X, Y are numpy arrays
+        Y = list(Y)
+
+        # get sift features
+        sift_features = self.get_sift_features(images=X)
+        # train kmeans
+        self.train_kmeans(sift_features=sift_features)
+        # get histogram
+        hist_features = self.histogram_from_sift(sift_features=sift_features)
+        # train svm
+        if svm_type == 'svc':
+            class_weight = 'balanced' if len(X) > 10 else None
+            self.svm_classifier = svm.SVC(class_weight=class_weight)
+        elif svm_type == 'one_class_svm':
+            self.svm_classifier = svm.OneClassSVM(gamma=gamma, nu=nu)
+        self.svm_classifier.fit(hist_features, Y)
+
+    def predict(self, X):
+        """
+        test the classifier
+        args:
+            X: a list of images
+        """
+        assert self.svm_classifier is not None
+        X = list(X)  # in case X is numpy array
+
+        # preprocess the images
+        sift_features = self.get_sift_features(images=X)  # get sift features
+        hist_features = self.histogram_from_sift(sift_features=sift_features)  # get histogram
+
+        return self.svm_classifier.predict(hist_features)
+    
+    def save(self, save_path):
+        """
+        save the classifier to disk
+        """
+        with open(save_path, 'wb') as f:
+            pickle.dump(self, f)
+    
+    @classmethod
+    def load(cls, load_path):
+        """
+        init a classifier by loading it from disk
+        """
+        with open(load_path, 'rb') as f:
+            return pickle.load(f)
+    
+    def get_sift_features(self, images):
+        """
+        extract the SIFT features of a list of images
+        args:
+            images: a list of RGB/GrayScale images
+        return:
+            a list of SIFT features
+        """
+        keypoints = self.sift_detector.detect(images)
+        keypoints, descriptors = self.sift_detector.compute(images, keypoints)
+        return descriptors  # type: tuple
+    
+    def train_kmeans(self, sift_features):
+        """
+        train the kmeans classifier using the SIFT features
+        args:
+            sift_features: a list of SIFT features
+        """
+        # reshape the data
+        # each image has a different number of descriptors, we should gather 
+        # them together to train the clustering
+        sift_features=np.array(sift_features, dtype=object)
+        sift_features=np.concatenate(sift_features, axis=0).astype(np.float32)
```

Comment:
Suppose we have N images with D features each. Doesnt this operation flatten the N, D matrix to a (N*D,) matrix? Is that ok for kmeans?